lib/charms/opensearch/v0/opensearch_health.py

```diff
@@ -0,0 +1,149 @@
+
+from charms.opensearch.v0.constants_charm import (
+    ClusterHealthRed,
+    ClusterHealthYellow,
+    WaitingForBusyShards,
+    WaitingForSpecificBusyShards,
+)
+from charms.opensearch.v0.helper_charm import Status
+from charms.opensearch.v0.helper_cluster import ClusterState
+from charms.opensearch.v0.helper_databag import Scope
+from charms.opensearch.v0.opensearch_exceptions import OpenSearchHttpError
+from ops.model import BlockedStatus, WaitingStatus
+
+# The unique Charmhub library identifier, never change it
+LIBID = "93d2c27f38974a59b3bbe39fb27ac98d"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+logger = logging.getLogger(__name__)
+
+
+class HealthColors:
+    """Colors the clusters or a unit may have depending on their health."""
+
+    GREEN = "green"
+    YELLOW = "yellow"
+    YELLOW_TEMP = "yellow-temp"
+    RED = "red"
+
+
+class OpenSearchHealth:
+    """Class for managing OpenSearch statuses."""
+
+    def __init__(self, charm):
+        self._charm = charm
+        self._opensearch = self._charm.opensearch
+
+    def apply(
+        self,
+        wait_for_green_first: bool = False,
+        use_localhost: bool = True,
+        app: bool = True,
+    ) -> str:
+        """Fetch cluster health and set it on the app status."""
+        host = self._charm.unit_ip if use_localhost else None
+        status = self._fetch_status(host, wait_for_green_first)
+
+        if app:
+            self.apply_for_app(status)
+        else:
+            self.apply_for_unit(status)
+
+        return status
+
+    def apply_for_app(self, status: str):
+        """Cluster wide / app status."""
+        if not self._charm.unit.is_leader():
+            # this is needed in case the leader is in an error state and doesn't
+            # report the status itself
+            self._charm.peers_data.put(Scope.UNIT, "health", status)
+            return status
+
+        if status == HealthColors.GREEN:
+            # green: cluster healthy
+            self._charm.status.clear(ClusterHealthRed, app=True)
+            self._charm.status.clear(ClusterHealthYellow, app=True)
+            self._charm.status.clear(WaitingForBusyShards, app=True)
+        elif status == HealthColors.RED:
+            # health RED, some primary shards are unassigned
+            self._charm.app.status = BlockedStatus(ClusterHealthRed)
+            return
+        else:
+            # cluster health is yellow, either a temporary state where shards are moving
+            # around (relocating / initializing) - or a permanent state where some replica
+            # shards are unassigned.
+            self._charm.app.status = (
+                BlockedStatus(ClusterHealthYellow)
+                if status == HealthColors.YELLOW
+                else WaitingStatus(WaitingForBusyShards)
```

Comment:
feel free to ignore: If we have a cascade of ifs, for better readability I prefer to have a cascade for all cases (not inline ifs for some cases).